blockchain/storagev2/leveldb/leveldb_test.go

```diff
@@ -0,0 +1,256 @@
+		_, err := rand.Read(input)
+
+		require.NoError(t, err)
+
+		tx.ComputeHash()
+
+		txs[i] = tx
+	}
+
+	return txs
+}
+
+func generateBlock(t *testing.T, num uint64) *types.FullBlock {
+	t.Helper()
+
+	transactionsCount := 2500
+	status := types.ReceiptSuccess
+	addr1 := types.StringToAddress("17878aa")
+	addr2 := types.StringToAddress("2bf5653")
+	b := &types.FullBlock{
+		Block: &types.Block{
+			Header: &types.Header{
+				Number:    num,
+				ExtraData: make([]byte, 32),
+				Hash:      types.ZeroHash,
+			},
+			Transactions: generateTxs(t, 0, transactionsCount, addr1, &addr2),
+			Uncles:       blockchain.NewTestHeaders(10),
+		},
+		Receipts: make([]*types.Receipt, transactionsCount),
+	}
+
+	logs := make([]*types.Log, 10)
+
+	for i := 0; i < 10; i++ {
+		logs[i] = &types.Log{
+			Address: addr1,
+			Topics:  []types.Hash{types.StringToHash("t1"), types.StringToHash("t2"), types.StringToHash("t3")},
+			Data:    []byte{0xaa, 0xbb, 0xcc, 0xdd, 0xbb, 0xaa, 0x01, 0x012},
+		}
+	}
+
+	for i := 0; i < len(b.Block.Transactions); i++ {
+		b.Receipts[i] = &types.Receipt{
+			TxHash:            b.Block.Transactions[i].Hash(),
+			Root:              types.StringToHash("mockhashstring"),
+			TransactionType:   types.LegacyTxType,
+			GasUsed:           uint64(100000),
+			Status:            &status,
+			Logs:              logs,
+			CumulativeGasUsed: uint64(100000),
+			ContractAddress:   &types.Address{0xaa, 0xbb, 0xcc, 0xdd, 0xab, 0xac},
+		}
+	}
+
+	for i := 0; i < 5; i++ {
+		b.Receipts[i].LogsBloom = types.CreateBloom(b.Receipts)
+	}
+
+	return b
+}
+
+func newStorageP(t *testing.T) (*storagev2.Storage, func(), string) {
+	t.Helper()
+
+	p, err := os.MkdirTemp("", "leveldbV2-test")
+	require.NoError(t, err)
+
+	require.NoError(t, os.MkdirAll(p, 0755))
+
+	s, err := NewLevelDBStorage(p, hclog.NewNullLogger())
+	require.NoError(t, err)
+
+	closeFn := func() {
+		require.NoError(t, s.Close())
+
+		if err := s.Close(); err != nil {
+			t.Fatal(err)
+		}
+
+		require.NoError(t, os.RemoveAll(p))
+	}
+
+	return s, closeFn, p
+}
+
+func countLdbFilesInPath(path string) int {
+	pattern := filepath.Join(path, "*.ldb")
+
+	files, err := filepath.Glob(pattern)
+	if err != nil {
+		return -1
+	}
+
+	return len(files)
+}
+
+func generateBlocks(t *testing.T, count int, ch chan *types.FullBlock, ctx context.Context) {
+	t.Helper()
+
+	ticker := time.NewTicker(100 * time.Millisecond)
+
+	for i := 1; i <= count; i++ {
+		b := generateBlock(t, uint64(i))
+		select {
+		case <-ctx.Done():
+			close(ch)
+			ticker.Stop()
+
+			return
+		case <-ticker.C:
+			ch <- b
+		}
+	}
+}
+
+func dirSize(t *testing.T, path string) int64 {
+	t.Helper()
+
+	var size int64
+
+	err := filepath.Walk(path, func(_ string, info os.FileInfo, err error) error {
+		if err != nil {
+			t.Fail()
+		}
+
+		if !info.IsDir() {
+			size += info.Size()
+		}
+
+		return err
+	})
+	if err != nil {
+		t.Log(err)
+	}
+
+	return size
+}
+
+func TestWriteFullBlock(t *testing.T) {
```

Comment:
Just a thought, but since these tests are the same for `mdbx` and `leveldb`, can we make a helper function in let's say `helper_test.go`, which will take `storagev2.Storage` interface and run the test logic on it? It will be something like this:

In `leveldb_test.go` and `mdbx_test.go` you will have the same test (`TestWriteFullBlock`), which will create their respective storages and just pass them to `helper_test.go` function:

```go
func TestWriteFullBlock(t *testing.T) {
    s, _, path := newStorageP(t)
    defer s.Close()
    
    helperTest.TestWriteFullBlock(t, s)
}
```

In `helper_test.go`:
```gp
func TestWriteFullBlock(t *testing.T, storage storagev2.Storage) {
      // test logic
      ...
}
```

The same would apply for the performance tests.

Reply:
There are some minor differences in tests itself which prevents from implementing this. At the end of for loop there is calculation of directory size which differs in mdbx and leveldb tests.

mdbx
size := dbSize(t, path)
t.Logf("writing block %d", i)
t.Logf("\tdir size %d MBs", size/1_000_000)

leveldb
size := dirSize(t, path)
t.Logf("writing block %d", i)
t.Logf("\tldb file count: %d", countLdbFilesInPath(path))
t.Logf("\tdir size %d MBs", size/1_000_000)